FP: add KZG point-evaluation preimage-oracle type

protolambda opened the pull request:
**Description**

The idea of this new preimage type is that we substitute the L2 KZG point-evaluation precompile, so we do not run billions of cryptography instructions in MIPS when we don't really need to.

This is a draft, covering:
- add the preimage key type
- add the preimage hint
- add the smart-contract code to load the preimage data
- add the precompile implementation
- add some TODOs on how to finish this work

What remains (now implemented in this PR):
- [x] Tiny op-geth change to be able to substitute in the precompile, conditionally on MIPS, or conditionally on op-program. Technically we can avoid this by just writing to the global precompiles map, and making this whole thing conditional on the MIPS build target (or custom cannon build target, if we wanted to). Later we would also want to support RISC-V and maybe WASM. It would be cleaner to just make it conditional on the op-program being the user of the EVM.
- [x] Change the op-program host to "prepare" the preimage: parse the hint, call the actual verification function, and make the yes/no result the preimage by turning it into a byte 1 for success, 0 for error
- [x] Change the challenger to support this, if needed. Technically it's not relying on external data, so the op-program host should be able to do all the work. Unless the challenger harcodes preimage types somewhere. Edit: the challenger does need to know to call this new solidity function with the exact data that is already in the hint.
- [x] Specs PR to add the new preimage-key type, and clarify how the fault-proof program can use this to support the KZG precompile functionality.


## Challenger Interaction

A new preimage [type `6`](https://github.com/ethereum-optimism/specs/blob/main/specs/experimental/fault-proof/index.md#type-6-global-eip-4844-point-evaluation-precompile-key) is used for the KZG precompile. The op-challenger will be able to fetch the input preimage from its local oracle to be forwarded to the `PreimageOracle`.
This gives us the following interaction:

#### Local
1. FPP EVM requests KZG precompile results for `x`, where `x` is a hash of the precompile input.
2. FPP hints `preimage(x)`.
3. Host prepares the hint and preimage result. Both are stored in the local oracle.

##### Onchain - Preparing a VM.step for a precompile computation
1. The challenger fetches `preimage(x)` and calls `PreimageOracle.loadKZGPointEvaluationPreimage` with the preimage as input.
2. Oracle contract computes the precompile result and stores it.
3. The fault proof program is then able to retrieve the result using `x`.

protolambda committed:
FP: add KZG point-evaluation preimage-oracle type, to process L2 oracle call with

protolambda commented on op-program/client/l2/engineapi/kzg_precompile.go:
```diff
@@ -0,0 +1,122 @@
+	}
+	// ------------------------------------------------------------------
+
+	return common.Hex2Bytes(blobPrecompileReturnValue), nil
```
Argh, geth uses the footgun (correctly at least) in consensus critical code :/. I copied it over (see copyright notice comment on top of file), but probably worth changing to the safe function.

tynes commented:
specs pr if this is srs plz 👉👈

Inphi committed:
op-program: Integrate kzg precompile oracle

Inphi committed:
fix PreimageOracle.sol

Inphi committed:
store kzg precompile input preimage

Inphi committed:
update bindings and go.mod

Inphi committed:
fix go.mod

Inphi committed:
resolve TODOs

Inphi committed:
s/Hex2Bytes/FromHex

Inphi committed:
fix typo

clabby commented on packages/contracts-bedrock/src/cannon/PreimageOracle.sol:
```diff
@@ -329,6 +329,45 @@ contract PreimageOracle is IPreimageOracle {
+        // the part offset is always 0
+        preimagePartOk[key][0] = true;
+        preimageParts[key][0] = part;
+        preimageLengths[key] = 32;
```
Are we intentionally setting the preimage size to `32` rather than `1` here?

Inphi replied:
Good callout. It's `32` since that's the data length of preimage chunks read by MIPS.sol.
Though this is a bit confusing as it deviates from the preimage length of the local oracle. I'll look into changing this.

clabby reviewed (approved):
Contract changes LGTM, pending the preimage size question.

refcell committed:
Merge branch 'develop' into kzg-point-evaluation-oracle

refcell commented on op-program/client/l2/engine_backend_test.go:
```diff
@@ -257,10 +287,11 @@ func createBlock(t *testing.T, chain *OracleBackedL2Chain) *types.Block {
 		rawTx := &types.DynamicFeeTx{
 			ChainID:   config.ChainID,
 			Nonce:     nonce,
-			To:        &targetAddress,
+			To:        target,
 			GasTipCap: big.NewInt(0),
 			GasFeeCap: parent.BaseFee(),
-			Gas:       21_000,
+			Gas:       210_000,
```
Was this intentional?

Inphi replied:
Yeah. The kzg precompile needs more gas than intrinsic.

refcell reviewed (commented):
1 comment

Inphi committed:
ctb: set kzg preimage length to 1

Inphi committed:
op-challenger: Load kzg point evaluations to PreimageOracle (#9497)

* op-challenger: load KZG point evaluation into oracle

Also: increase max game depth for devnet.
This is a temporary solution to avoid flakes in e2e tests that use the kzg precompile.
The current execution trace max step of 2**31 steps isn't sufficient.

Inphi commented on op-e2e/e2eutils/disputegame/output_game_helper.go:
```diff
@@ -558,7 +559,13 @@ func (g *OutputGameHelper) uploadPreimage(ctx context.Context, data *types.Preim
 	boundOracle, err := bindings.NewPreimageOracle(oracle.Addr(), g.client)
 	g.require.NoError(err)
-	tx, err := boundOracle.LoadKeccak256PreimagePart(g.opts, new(big.Int).SetUint64(uint64(data.OracleOffset)), data.GetPreimageWithoutSize())
+	var tx *gethtypes.Transaction
+	switch data.OracleKey[0] {
+	case byte(preimage.KZGPointEvaluationKeyType):
+		tx, err = boundOracle.LoadKZGPointEvaluationPreimage(g.opts, data.GetPreimageWithoutSize())
+	default:
+		tx, err = boundOracle.LoadKeccak256PreimagePart(g.opts, new(big.Int).SetUint64(uint64(data.OracleOffset)), data.GetPreimageWithoutSize())
+	}
```
I'll follow up on this on another PR. We need to refactor the IsLocal/Global concept now that we have blob and precompile preimages.
